### PR TITLE
[VL] Make bloom_filter_agg fall back when might_contain is not transformable

### PR DESCRIPTION
**Pull request by zhli1142015**
## What changes were proposed in this pull request?

Fallback bloom_filter_agg when might_contain fall back, either because of might_contain  self or other expressions in same operator. This fixes below error:
```
java.io.IOException: Unexpected Bloom filter version number (16777217)
  at org.apache.spark.util.sketch.BloomFilterImpl.readFrom0(BloomFilterImpl.java:256)
  at org.apache.spark.util.sketch.BloomFilterImpl.readFrom(BloomFilterImpl.java:265)
  at org.apache.spark.util.sketch.BloomFilter.readFrom(BloomFilter.java:178)
  at org.apache.spark.sql.catalyst.expressions.BloomFilterMightContain.deserialize(BloomFilterMightContain.scala:111)

```
## How was this patch tested?

UT.

### CODE REVIEW

**Comment by zhli1142015** [comment]
@JkSelf , @PHILO-HE , @jinchengchenghh , coudl you help to review this?
Thanks.

File: gluten-core/src/main/scala/io/glutenproject/extension/columnar/TransformHintRule.scala [hunk 1]
```diff
@@ -714,6 +714,13 @@ case class AddTransformHintRule() extends Rule[SparkPlan] {
           s"${e.getMessage}, original sparkplan is " +
             s"${plan.getClass}(${plan.children.toList.map(_.getClass)})")
     }
+
+    if (TransformHints.isAlreadyTagged(plan) && TransformHints.isNotTransformable(plan)) {
```
Comment (PHILO-HE) [inline gluten-core/src/main/scala/io/glutenproject/extension/columnar/TransformHintRule.scala:718]: Maybe, we can move this part into a dedicated rule after  `AddTransformHintRule` (see [link](https://github.com/oap-project/gluten/blob/9d07d48e2f019bd1d3b1e36798c32f7cce62915c/gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala#L817)).

File: shims/spark34/src/main/scala/io/glutenproject/sql/shims/spark34/Spark34Shims.scala [hunk 2]
```diff
@@ -130,6 +130,33 @@ class Spark34Shims extends SparkShims {
       @transient locations: Array[String] = Array.empty): PartitionedFile =
     PartitionedFile(partitionValues, SparkPath.fromPathString(filePath), start, length, locations)
 
+  override def handleBloomFilterFallback(plan: SparkPlan)(fun: SparkPlan => Unit): Unit = {
```
Comment (PHILO-HE) [inline shims/spark34/src/main/scala/io/glutenproject/sql/shims/spark34/Spark34Shims.scala:133]: This method looks repeated. If so, maybe, we can just move this method into a common place. But just move the check for `BloomFilterAggregate` into shims, as I note it doesn't exist in spark3.2.

Spark 3.3/3.4:
```
def hasBloomFilterAggregate(agg): Boolean = {
  agg.aggregateExpressions.exists( expr.aggregateFunction.isInstanceOf[BloomFilterAggregate])
}
```
Spark 3.2:
```
def hasBloomFilterAggregate(agg): Boolean = {
   false
}
```

**Review by PHILO-HE — COMMENTED** [review]
Thanks for your work! Just some suggestions.

**Comment by PHILO-HE** [comment]
Also note this doc: https://github.com/oap-project/gluten/blob/main/docs/velox-backend-limitations.md#runtime-bloomfilter. With this fix, we need to document the behavior into fallback section. Thanks!

File: gluten-core/src/main/scala/io/glutenproject/extension/columnar/TransformHintRule.scala (hunk 1 above)
Comment (zhli1142015) [reply gluten-core/src/main/scala/io/glutenproject/extension/columnar/TransformHintRule.scala]: Thanks, moved to a separate rule.

File: shims/spark34/src/main/scala/io/glutenproject/sql/shims/spark34/Spark34Shims.scala (hunk 2 above)
Comment (zhli1142015) [reply shims/spark34/src/main/scala/io/glutenproject/sql/shims/spark34/Spark34Shims.scala]: Updated, thanks,

File: gluten-core/src/main/scala/io/glutenproject/extension/columnar/TransformHintRule.scala [hunk 3]
```diff
@@ -270,6 +270,44 @@ case class FallbackEmptySchemaRelation() extends Rule[SparkPlan] {
   }
 }
 
+/**
+ * Velox BloomFilter's implementation is different from Spark's. So if might_contain falls back, we
+ * need fall back related bloom filter agg.
+ */
+case class FallbackBloomFilterAggIfNeeded() extends Rule[SparkPlan] {
+  override def apply(plan: SparkPlan): SparkPlan = plan.transformDown {
```
Comment (PHILO-HE) [inline gluten-core/src/main/scala/io/glutenproject/extension/columnar/TransformHintRule.scala:278]: Looks good! Maybe, we can skip the handling and just return the original plan if `spark.gluten.sql.native.bloomFilter=false`. Right?

Comment (zhli1142015) [reply gluten-core/src/main/scala/io/glutenproject/extension/columnar/TransformHintRule.scala]: Make sense, thanks.

**Review by PHILO-HE — APPROVED** [review]
Looks good! Thanks!

**Review by JkSelf — APPROVED** [review]
LGTM. Thanks @zhli1142015 @PHILO-HE .

**Pull request merged**